sys/kern/signal.c

```diff
@@ -329,70 +349,50 @@ void sig_kill(proc_t *p, ksiginfo_t *ksi) {
     return;
 
   thread_t *td = p->p_thread;
+  bool ignored = sig_ignored(p->p_sigactions, sig);
 
-  sig_t handler = p->p_sigactions[sig].sa_handler;
-  bool continued = sig == SIGCONT || sig == SIGKILL;
-
-  /* If the signal is ignored, don't even bother posting it,
-   * unless it's waking up a stopped process. */
-  if (p->p_state == PS_STOPPED && continued) {
-    p->p_state = PS_NORMAL;
-    p->p_flags |= PF_STATE_CHANGED;
-    WITH_PROC_LOCK(p->p_parent) {
-      proc_wakeup_parent(p->p_parent);
-    }
-  } else if (handler == SIG_IGN ||
-             (defact(sig) == SA_IGNORE && handler == SIG_DFL)) {
+  if (ignored && !sigprop_cont(sig))
     return;
-  }
 
   /* Theoretically, we should hold all_proc_mtx since sig_ignore_ttystop()
    * reads pg_jobc, but the race here is harmless. */
   if (sig_ignore_ttystop(p, sig))
     return;
 
-  /* If stopping or continuing,
+  /* If sending a stop or continue signal,
    * remove pending signals with the opposite effect. */
-  if (defact(sig) == SA_STOP)
+  if (defact_stop(sig)) {
     sigpend_get(&td->td_sigpend, SIGCONT, NULL);
+  } else if (sigprop_cont(sig)) {
+    sigpend_delete_set(&td->td_sigpend, &stopmask);
+    if (p->p_state == PS_STOPPED) {
```

Comment:
Maybe introduce `proc_continue` counterpart to `proc_stop` ?

Reply:
Good idea. Done.